src/net/server/mod.rs

```diff
@@ -77,24 +81,110 @@ impl From<axum::http::header::ToStrError> for MpcHelperServerError {
     }
 }
 
+impl From<axum::extract::rejection::PathRejection> for MpcHelperServerError {
+    fn from(err: axum::extract::rejection::PathRejection) -> Self {
+        Self::BadPathString(err.into())
+    }
+}
+
+/// [`From`] implementation for [`MpcServerError::SendError`].
+/// first call `to_string` so as to drop `T` from the [`MpcServerError`]
+impl<T> From<mpsc::error::SendError<T>> for MpcHelperServerError {
+    fn from(err: mpsc::error::SendError<T>) -> Self {
+        Self::SendError(err.to_string().into())
+    }
+}
+
+/// [`From`] implementation for [`MpcServerError::SendError`].
+/// first call `to_string` to as to drop `T` from the [`MpcServerError`]
+impl<T> From<tokio_util::sync::PollSendError<T>> for MpcHelperServerError {
+    fn from(err: tokio_util::sync::PollSendError<T>) -> Self {
+        Self::SendError(err.to_string().into())
+    }
+}
+
 impl IntoResponse for MpcHelperServerError {
     fn into_response(self) -> Response {
         let status_code = match &self {
             Self::BadQueryString(_) | Self::BadPathString(_) | Self::MissingHeader(_) => {
                 StatusCode::UNPROCESSABLE_ENTITY
             }
-            Self::SerdeError(_) | Self::InvalidHeader(_) | Self::WrongBodyLen { .. } => {
-                StatusCode::BAD_REQUEST
-            }
-            Self::HyperError(_) | Self::SendError(_) | Self::BodyAlreadyExtracted(_) => {
-                StatusCode::INTERNAL_SERVER_ERROR
-            }
+            Self::SerdeError(_)
+            | Self::InvalidHeader(_)
+            | Self::WrongBodyLen { .. }
+            | Self::OutOfOrder { .. } => StatusCode::BAD_REQUEST,
+            Self::HyperError(_)
+            | Self::SendError(_)
+            | Self::BodyAlreadyExtracted(_)
+            | Self::MissingExtension(_) => StatusCode::INTERNAL_SERVER_ERROR,
         };
 
         (status_code, self.to_string()).into_response()
     }
 }
 
+/// Provides a mapping of [`QueryId`]s to senders that forward data to a [`Network`]. Every time a
+/// new query is started, a [`Network`] is created to handle communication for that query. When the
+/// server receives a request, it must know which [`Network`] to forward that request to, so it
+/// holds this mapping to accomplish that.
+///
+/// Is shareable by `clone()`ing.
+#[derive(Clone)]
+pub struct MessageSendMap {
+    senders: Arc<Mutex<HashMap<QueryId, mpsc::Sender<MessageChunks>>>>,
```

Comment:
is there a way we can store `T: Network` instead? Do you need to share network between this struct and anything else? `Network` is sync, so any reference to it will be `Send` and therefore thread safe

Reply:
we probably could, but we'd have to change the trait to expose the opposite end of the `recv_stream`, no?

Reply:
you're right `T: Network` won't work, but can we store the actual struct that implements `Network` here? It should expose methods to send data. I guess I just don't like the fact that we assume and expose the actual implementation for `Network` trait - it should be the internal business of `Network` implementation how to manage connection and communication. Structs outside it ideally just use public API of that struct to send data to it

Reply:
I agree that makes sense. The only thing I'm thinking about is the `&'a peers` field in the struct. How would I handle that ref within `MessageSendMap`?

Reply:
have you tried something like this?

```rust
struct QueryHandler<'a> {
    inner: Arc<Mutex<HashMap<QueryId, &'a MyConfig<'a>>>>,
}

You can also clone - we need to be careful with it and that's why I think making `PeerDiscovery` return a reference was the right thing to do. but if it is a blocker, I think we can live with it